sklearn/neural_network/_stochastic_optimizers.py

```diff
@@ -0,0 +1,265 @@
+
+        Parameters
+        ----------
+        msg : str
+            Message passed in for verbose output
+        verbose : bool
+            Print message to stdin if True
+
+        Returns
+        -------
+        is_stopping : bool
+            True if traning needs to stop
+        """
+        if verbose:
+            print(msg + " Stopping.")
+        return True
+
+
+class SGDOptimizer(BaseOptimizer):
+    """Stochastic gradient descent optimizer with momentum
+
+    Parameters
+    ----------
+    params : list, length = len(coefs_) + len(intercepts_)
+        The list containing coefs_ and intercepts_ concatenated in MLP model.
+        Used for initializing velocities and updating params
+
+    learning_rate_init : float, optional, default 0.1
+        The initial learning rate used. It controls the step-size in updating
+        the weights
+
+    lr_schedule : {'constant', 'adaptive', 'invscaling'}, default 'constant'
+        Learning rate schedule for weight updates.
+
+        -'constant', is a constant learning rate given by
+         'learning_rate_init'.
+
+        -'invscaling' gradually decreases the learning rate 'learning_rate_' at
+          each time step 't' using an inverse scaling exponent of 'power_t'.
+          learning_rate_ = learning_rate_init / pow(t, power_t)
+
+        -'adaptive', keeps the learning rate constant to
+         'learning_rate_init' as long as the training keeps decreasing.
+         Each time 2 consecitive epochs fail to decrease the training loss by
+         tol, or fail to increase validation score by tol if 'early_stopping'
+         is on, the current learning rate is divided by 5.
+
+    momentum : float, optional, default 0.9
+        Value of momentum used, must be larger than or equal to 0
+
+    nesterov : bool, optional, default True
+        Whether to use nesterov's momentum or not. Use nesterov's if True
+
+    Attributes
+    ----------
+    learning_rate : float
+        the current learning rate
+
+    velocities : list, length = len(params)
+        velocities that are used to update params
+    """
+
+    def __init__(self, params, learning_rate_init=0.1, lr_schedule='constant',
+                 momentum=0.9, nesterov=True, power_t=0.5):
+        super(SGDOptimizer, self).__init__(params, learning_rate_init)
+
+        self.lr_schedule = lr_schedule
+        self.momentum = momentum
+        self.nesterov = nesterov
+        self.power_t = power_t
+        self.velocities = [np.zeros_like(param) for param in params]
+
+    def iteration_ends(self, time_step):
+        """Perform updates to learning rate and potential other states at the
+        end of an iteration
+
+        Parameters
+        ----------
+        time_step : int
+            number of training samples trained on so far, used to update
+            learning rate for 'invscaling'
+        """
+        if self.lr_schedule == 'invscaling':
+            self.learning_rate = (self.learning_rate_init /
```

Comment:
just to be sure, can you convert to float? È.g. `float(self.learning_rate_init)` to avoid inexpicable zeros if anybody ever sets the learning rate to an integer and the denominator also happens to be an integer?